libpoolprotocols/PoolManager.cpp

```diff
@@ -67,19 +67,17 @@ PoolManager::PoolManager(boost::asio::io_service& io_service, PoolClient* client
 
 	});
 
-	p_client->onWorkReceived([&](WorkPackage const& wp, bool check_for_duplicates)
-	{
-		if (check_for_duplicates) {
-			for (auto h : m_headers)
-				if (h == wp.header)
-				{
-					cwarn << EthYellow "Duplicate job #" << wp.header.abridged() << " discarded" EthReset;
-					return;
-				}
-		}
-		m_headers.push_back(wp.header);
+    p_client->onWorkReceived([&](WorkPackage const& wp) {
+        for (auto h : m_headers)
+            if (h == wp.header)
```

Comment:
Are we sure this equality check only on header is enough ?

According to farm setWork the job **is not** switched if this is true

```
if (_wp.header == m_work.header && _wp.startNonce == m_work.startNonce)
```
Same header with different startNonce is effectively a new work.
So at least we should settle only one check. Either in PoolManager or in Farm

And my other doubt is related to the fact we should also check for changed boundary.

If Nicehash sends a change in diff **and** resends the same job header ... isn't it a new job ?
How wp.boundary affects found solution ?

Reply:
I'm not sure why that IF statement is even there? We now will never call setWork unless the protocol has notified us of a new job. We no longer call setWork after stratum2 diff and extraNonce notifications.

Let me fix that. Yep... not needed. Pushing the change.

Reply:
Indeed, if they were going to catch the duplicate job created by the bugs in notify, whoever did it should also have checked the diff, in that IF statement.

Reply:
So it ends up only checking duplicate headers ?
R u sure ? All other members like extranonce and diff are not relevant ?

Reply:
Yes, that is correct. extranonce and start nonce can be updated at any time, but the target and startnonce parameters passed to both the opencl and nvidia kernels are only updated with there is a job hash mismatch. So, like the spec. says, we can update extranonce and diff (target) at any time, but these new values will only be picked up by the kernels at job switch.

Reply:
Finally stratum2 works as it should. No more false duplicates! Tested at daggerhashimoto.usa.nicehash.com through many dificulty and chain switches. You get some idea of which blockchain you are mining by looking at the epoch number changes.